File: TeamCode/src/main/java/virtualRobot/LogicThread.java
package virtualRobot;


import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import virtualRobot.commands.Command;
import virtualRobot.commands.Rotate;
import virtualRobot.commands.SpawnNewThread;
import virtualRobot.commands.Translate;

/**
 * Created by ethachu19 on 3/31/17
 *
 * LogicThread is the main working thread that runs procedurally
 * In realRun use runCommand to run commands
 * Do not use Command.changeRobotState
 *
 * HAHA TAKE THAT SUCKERS
 */
public abstract class LogicThread implements Runnable {
    protected List<Thread> children = new ArrayList<>(); //contains a list of threads created under this logic Thread using spawn new thread
    protected SallyJoeBot robot = Command.ROBOT;
    protected ConcurrentHashMap<String, Boolean> monitorData = new ConcurrentHashMap<>();
    private ConcurrentHashMap<Condition, LogicThread> interrupts = new ConcurrentHashMap<>();
    protected volatile Command lastRunCommand = null;
    private List<MonitorThread> monitorThreads = Collections.synchronizedList(new ArrayList<MonitorThread>());
    private volatile long startTime, elapsedTime = 0;

    private enum ThreadState{
        NOT_RUNNING,
        RUNNING,
        PAUSE
    };
    private volatile ThreadState state;

    //The thread to check all monitorThreads and put data in HashMap
    private Thread interruptHandler = new Thread() {

        public LogicThread parent = null;

        public Thread setParent(LogicThread logicThread) {
            parent = logicThread;
            return this;
        }

        public void run() {
//            parent.startTime = System.currentTimeMillis();
            boolean isInterrupted = false;
            while (!isInterrupted) {
                for (MonitorThread m : parent.monitorThreads) {
                    parent.monitorData.put(m.getClass().getName(), !m.getStatus());
                }

                for (Map.Entry<Condition, LogicThread> entry : parent.interrupts.entrySet()) {
                    if (entry.getKey().isConditionMet()) {
                        parent.pauseParent();
                        try {
                            parent.wait();
                        } catch (InterruptedException e) {
                            isInterrupted = true;
                            return;
                        }
                        if (parent.lastRunCommand != null)
                            parent.lastRunCommand.stopCommand(); //stops currently running command
                        entry.getValue().state = ThreadState.RUNNING;
                        try {
                            entry.getValue().realRun();
                        } catch (InterruptedException e) {
                            isInterrupted = true;
                        }
                        entry.getValue().killChildren();
                        entry.getValue().state = ThreadState.NOT_RUNNING;
                        parent.resumeParent();
                        parent.notify();
                    }
                }
//                parent.elapsedTime = System.currentTimeMillis() - startTime;
                isInterrupted = Thread.currentThread().isInterrupted();
                if (isInterrupted)
                    break;

                try {
                    Thread.sleep(1);
                } catch (InterruptedException ex) {
                    isInterrupted = true;
                    break;
                }
            }
        }
    }.setParent(this);

    @Override
    public void run(){
        state = ThreadState.RUNNING;
        interruptHandler.start();
        try {
            realRun();
        } catch (InterruptedException e) {
        }
        killChildren();
        interruptHandler.interrupt();
        state = ThreadState.NOT_RUNNING;
    }

    /**
     * Runs command with certain conditions to check type of command
     *
     * @param c Command to run
     * @return Whether command stopped by interrupt
     */
    protected synchronized boolean runCommand(Command c) throws InterruptedException{
        boolean isInterrupted = false;
        boolean stopByIH = false;
        lastRunCommand = c;

        if (state == ThreadState.PAUSE){
            if (waitForNotify())
                return true;
        }

        if (c instanceof SpawnNewThread) { //Add all children thread to list to kill later
            children.addAll(((SpawnNewThread)c).getThreads());
        }

        try {
            isInterrupted  = c.changeRobotState(); //Actually run the command
        } catch (InterruptedException e) {
            isInterrupted = true;
        }

        if (state == ThreadState.PAUSE && !isInterrupted) {
            stopByIH = true;
            if (c instanceof Translate) {
                Translate t = (Translate) c;
                if (t.getDirection().getCode() % 4 == 0) { // Went forward or backward
                    t.setTarget(t.getTarget() - Math.abs(robot.getLFEncoder().getValue() + robot.getRFEncoder().getValue() + robot.getLBEncoder().getValue() + robot.getRBEncoder().getValue()) / 4);
                } else if (t.getDirection().getCode() % 2 == 0) { //Went left or right
                    t.setTarget(t.getTarget() - Math.abs(robot.getLFEncoder().getValue() + robot.getLBEncoder().getValue()) / 2);
                } else {
                    int count = 0;
                    double avg = 0;
                    for (int i = 0; i < 4; i++) { //add distance based off only the motors that moved
                        if (t.getMultiplier()[i] != 0) {
                            count++;
                            switch (i) {
                                case 0:
                                    avg += robot.getLFEncoder().getValue();
                                    break;
                                case 1:
                                    avg += robot.getRFEncoder().getValue();
                                    break;
                                case 2:
                                    avg += robot.getLBEncoder().getValue();
                                    break;
                                case 3:
                                    avg += robot.getRBEncoder().getValue();
                                    break;
                            }
                        }
                    }
                    avg = Math.abs(avg) / count; //average the count
                    t.setTarget(avg / Math.sqrt(2)); //reduce it by multiplier
                }
            }
        }

        if (state == ThreadState.PAUSE){
            if (waitForNotify())
                return true;
        }

        if (stopByIH) //Resuming after stopped by Interrupt Handler, call command again and resume Translate where it left off
            try {
                if (c instanceof Translate) {
                    Translate t = (Translate) c;
                    isInterrupted = t.changeRobotState();
                } else {
                    isInterrupted = c.changeRobotState();
                }
            } catch (InterruptedException e) {
                isInterrupted = true;
            }
        return isInterrupted;
    }

    private boolean waitForNotify() {
        synchronized (this) {
            while (state == ThreadState.PAUSE) {
                try {
                    this.wait();
                } catch (InterruptedException e) {
                    return true;
                }
            }
        }
        return false;
    }

    public synchronized void removeMonitor(Class<?> object) {
        List<MonitorThread> remove = new ArrayList<>();
        for (MonitorThread mt : monitorThreads) {
            if (object.isInstance(mt))
                remove.add(mt);
        }
        for (MonitorThread mt : remove) {
            monitorThreads.remove(mt);
        }
    }

    /**
     * Method with real commands and running commands
     */
    protected abstract void realRun() throws InterruptedException;


    /**
     *
     * @param condition
     * @param action
     */
    public void attachInterrupt(Condition condition, LogicThread action) {
        interrupts.put(condition, action);
    }

    /**
     * Kills all children spawned by SpawnNewThread
     */
    public synchronized void killChildren() {
        for (Thread x : children) //Kill any threads made using spawn new thread
            if (x.isAlive())
                x.interrupt();
    }

    /**
     * Binds monitorThread to LogicThread so that it can reference it in code
     *
     * @param LogicThread The logic thread to
     * @param monitorThread monitorThread to be bound to LogicThread
     */
    public static void delegateMonitor(LogicThread LogicThread, MonitorThread monitorThread) {
        LogicThread.monitorThreads.add(monitorThread);
        LogicThread.monitorData.put(monitorThread.getClass().getName(), false);
    }

    /**
     * Put this logic thread into a pause state and notify other threads
     */
    protected synchronized void pauseParent() {
        synchronized (this) {
            state = ThreadState.PAUSE;
            notifyAll();
        }
    }

    /**
     * Resumes this logic thread and notifies other threads
     */
    protected synchronized void resumeParent() {
        synchronized (this) {
            state = ThreadState.RUNNING;
            notifyAll();
        }
    }
}

File: TeamCode/src/main/java/virtualRobot/LogicThread.kt
package virtualRobot


import java.util.ArrayList
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

import virtualRobot.commands.Command
import virtualRobot.commands.SpawnNewThread
import virtualRobot.commands.Translate
import virtualRobot.utils.BetterLog

/**
 * Created by ethachu19 on 3/31/17
 *
 * LogicThread is the main working thread that runs procedurally
 * In realRun use runCommand to run commands
 * Do not use Command.changeRobotState
 *
 * HAHA TAKE THAT SUCKERS
 */
abstract class LogicThread : Thread() {
    protected var children: MutableList<Thread> = ArrayList() //contains a list of threads created under this logic Thread using spawn new thread
    protected var robot = Command.ROBOT
    private val monitorData = ConcurrentHashMap<String, Boolean>()
    private val interrupts = ConcurrentHashMap<Condition, LogicThread>()
    @Volatile protected var lastRunCommand: Command? = null
    private val monitorThreads = Collections.synchronizedList(ArrayList<MonitorThread>())
    @Volatile private var startTime: Long = 0
    @Volatile private var elapsedTime: Long = 0

    protected var isPaused = AtomicBoolean(false)
    protected var shouldStartISR = true

    //The thread to check all monitorThreads and put data in HashMap and check for interrupts
    private val interruptHandler = object : Thread() {

        var parent: LogicThread? = null

        fun setParent(logicThread: LogicThread): Thread {
            parent = logicThread
            return this
        }

        override fun run() {
            parent!!.startTime = System.currentTimeMillis()
            var isInterrupted = false
            while (!isInterrupted) {
                for (m in parent!!.monitorThreads) {
                    parent!!.monitorData.put(m.javaClass.getName(), !m.status || parent!!.monitorData[m.javaClass.getName()])
                }

                for ((key, value) in parent!!.interrupts) {
                    if (key.isConditionMet) {
                        parent!!.isPaused.set(true)
                        if (parent!!.lastRunCommand != null)
                            parent!!.lastRunCommand!!.stopCommand() //stops currently running command
                        robot.stopMotors()
                        try {
                            value.realRun()
                        } catch (e: InterruptedException) {
                            isInterrupted = true
                        }

                        value.killChildren()
                        parent!!.lastRunCommand!!.resetStopCommand()
                        parent!!.isPaused.set(false)
                    }
                }
                parent!!.elapsedTime = System.currentTimeMillis() - startTime
                isInterrupted = Thread.currentThread().isInterrupted || isInterrupted
                if (isInterrupted)
                    break

                try {
                    Thread.sleep(1)
                } catch (ex: InterruptedException) {
                    isInterrupted = true
                    break
                }

            }
        }
    }.setParent(this)

    override fun run() {
        isPaused.set(false)
        addPresets()
        if (shouldStartISR) {
            interruptHandler.start()
        }
        try {
            realRun()
        } catch (e: InterruptedException) {
            BetterLog.d("INTERRUPTS", this.javaClass.getName() + " was interrupted")
        } finally {
            killChildren()
            killMonitorThreads()
            if (shouldStartISR)
                interruptHandler.interrupt()
            isPaused.set(false)
        }
    }

    /**
     * Runs command with certain conditions to check type of command
     *
     * @param c Command to run
     * @return Whether command stopped by interrupt
     */
    @Synchronized
    @Throws(InterruptedException::class)
    protected fun runCommand(c: Command) {
        var isInterrupted = false
        var stopByIH = false
        lastRunCommand = c

        if (isPaused.get()) {
            if (waitForNotify())
                throw InterruptedException("waitForNotify was interrupted")
        }

        if (c is SpawnNewThread) { //Add all children thread to list to kill later
            children.addAll(c.threads)
        }

        try {
            isInterrupted = c.changeRobotState() //Actually run the command
        } catch (e: InterruptedException) {
            isInterrupted = true
        }

        if (isInterrupted)
            throw InterruptedException(c.javaClass.getName() + " was interrupted")

        if (isPaused.get()) {
            stopByIH = true
            if (c is Translate) {
                if (c.direction.code % 4 == 0) { // Went forward or backward
                    c.target = c.target - Math.abs(robot.lfMotor.position + robot.rfMotor.position + robot.lbMotor.position + robot.rbMotor.position) / 4
                } else if (c.direction.code % 2 == 0) { //Went left or right
                    c.target = c.target - Math.abs(robot.lfMotor.position + robot.lbMotor.position) / 2
                } else {
                    var count = 0
                    var avg = 0.0
                    for (i in 0..3) { //add distance based off only the motors that moved
                        if (c.multiplier[i] != 0) {
                            count++
                            when (i) {
                                0 -> avg += robot.lfMotor.position.toDouble()
                                1 -> avg += robot.rfMotor.position.toDouble()
                                2 -> avg += robot.lbMotor.position.toDouble()
                                3 -> avg += robot.rbMotor.position.toDouble()
                            }
                        }
                    }
                    avg = Math.abs(avg) / count //average the count
                    c.target = avg / Math.sqrt(2.0) //reduce it by multiplier
                }
            }
        }

        if (isPaused.get()) {
            if (waitForNotify())
                throw InterruptedException("waitForNotify was interrupted")
        }

        if (stopByIH)
        //Resuming after stopped by Interrupt Handler, call command again and resume Translate where it left off
            try {
                if (c is Translate) {
                    isInterrupted = c.changeRobotState()
                } else {
                    isInterrupted = c.changeRobotState()
                }
            } catch (e: InterruptedException) {
                isInterrupted = true
            }

        if (isInterrupted)
            throw InterruptedException(c.javaClass.getName() + " was interrupted")
    }

    /**
     * Gets the data related to the monitorThread given
     *
     * @param obj Class to look for
     * @return Data
     */
    protected fun getMonitorData(obj: Class<out MonitorThread>): Boolean {
        val temp = monitorData[obj.name]
        monitorData.put(obj.name, false)
        return temp
    }

    private fun waitForNotify(): Boolean {
        synchronized(this) {
            while (isPaused.get()) {
                try {
                    //                    this.wait();
                    Thread.sleep(10)
                } catch (e: InterruptedException) {
                    return true
                }

            }
        }
        return false
    }

    /**
     * Removes all monitorThreads of class type object
     *
     * @param object class to remove
     */
    @Synchronized
    fun removeMonitor(`object`: Class<out MonitorThread>) {
        val remove = ArrayList<MonitorThread>()
        for (mt in monitorThreads) {
            if (`object`.isInstance(mt))
                remove.add(mt)
        }
        monitorData.remove(`object`.name)
        for (mt in remove) {
            mt.interrupt()
            monitorThreads.remove(mt)
        }
    }

    /**
     * Determines whether this thread or ISR is alive
     *
     * @return isAlive
     */
    fun allIsAlive(): Boolean {
        return this.isAlive || interruptHandler.isAlive
    }

    /**
     * Method with real commands and running commands
     */
    @Throws(InterruptedException::class)
    protected abstract fun realRun()

    /**
     * Used to delegate monitorThreads and attach interrupts
     */
    protected open fun addPresets() {}

    /**
     * Attaches intterupt to the ISR with a certain procedure
     *
     * @param condition
     * @param action
     */
    fun attachInterrupt(condition: Condition, action: LogicThread) {
        interrupts.put(condition, action)
    }

    /**
     * Kills all children spawned by SpawnNewThread
     */
    @Synchronized
    fun killChildren() {
        for (x in children)
        //Kill any threads made using spawn new thread
            if (x.isAlive)
                x.interrupt()
    }

    /**
     * Kills all monitorThreads associated with this LogicThread
     */
    @Synchronized private fun killMonitorThreads() {
        for (x in monitorThreads)
            if (x.isAlive)
                x.interrupt()
    }

    /**
     * Binds monitorThread to LogicThread so that it can reference it in code
     *
     * @param monitorThread monitorThread to be bound to LogicThread
     * @param runAsThread Whether or not monitorThread should be start as new Thread
     */
    @JvmOverloads
    fun delegateMonitor(monitorThread: MonitorThread, runAsThread: Boolean = false) {
        monitorThreads.add(monitorThread)
        monitorData.put(monitorThread.javaClass.getName(), false)
        if (runAsThread) {
            monitorThread.setThread(true)
            monitorThread.start()
        }
    }
}
/**
 * Binds monitorThread to LogicThread so that it can reference it in code without starting it as a Thread
 *
 * @param monitorThread monitorThread to be bound to LogicThread
 */
